FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+
+                    // step 7: update message id and message pop receipt to job lock entity
+                    // if enqueue concurrently, it is possible that
+                    // 1. one job sends message and updates entity, another job do nothing
+                    // 2. two jobs both send message, while only one job update entity successfully
+                    try
+                    {
+                        await _azureJobInfoTableClient.UpdateEntityAsync(
+                            jobLockEntity,
+                            jobLockEntity.ETag,
+                            cancellationToken: cancellationToken);
+                    }
+                    catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.UpdateEntityPreconditionFailedErrorCode))
+                    {
+                        _logger.LogWarning(ex, "Update job info entity conflicts.");
+                    }
+                }
+
+                jobInfos.Add(jobInfoEntity.ToJobInfo<TJobInfo>());
+            }
+
+            _logger.LogInformation($"Enqueue jobs '{string.Join(",", jobInfos.Select(jobInfo => jobInfo.Id).ToList())}' successfully.");
+
+            return jobInfos;
+        }
+
+        public async Task<JobInfo> DequeueAsync(byte queueType, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Start to dequeue.");
+
+            // step 1: receive message from message queue
+            TimeSpan? visibilityTimeOut = heartbeatTimeoutSec <= 0 ? null : TimeSpan.FromSeconds(heartbeatTimeoutSec);
+            var message = await _azureJobMessageQueueClient.ReceiveMessageAsync(visibilityTimeOut, cancellationToken);
+
+            if (message.Value == null)
+            {
+                _logger.LogWarning("Failed to receive message from queue, the queue message is null.");
+                return null;
+            }
+
+            var jobMessage = JsonConvert.DeserializeObject<JobMessage>(message.Value.Body.ToString());
+
+            if (jobMessage == null)
+            {
+                _logger.LogWarning("Failed to deserialize message, the deserialized job message is null.");
+                return null;
+            }
+
+            // step 2: get jobInfo entity to check job status, delete this message if job is already completed/failed/cancelled
+            // if status is running and CancelRequest is true, the job will be dequeued, and jobHosting will continue to handle it
+            var jobInfoEntityResponse = await _azureJobInfoTableClient.GetEntityAsync<JobInfoEntity>(
+                jobMessage.PartitionKey,
+                jobMessage.RowKey,
+                cancellationToken: cancellationToken);
+            var jobInfo = jobInfoEntityResponse.Value.ToJobInfo<TJobInfo>();
+
+            if (jobInfo.Status is JobStatus.Completed or JobStatus.Failed or JobStatus.Cancelled)
```

Comment:
Looks like this case should not happen?

Reply:
This case happens when updating the table entity successfully while failing to delete the message in `CompleteJobAsync`.

Reply:
Per our discussion, will throw an exception for it.